### PR TITLE
Fix the source of inlined trees after Erasure

### PR DESCRIPTION
**Pull request by nicolasstucki**
To ensure that the sources change, the trees need to be copied with the
new source rather than preserving it.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/Trees.scala [hunk 1]
```diff
@@ -981,6 +981,11 @@ object Trees {
       protected def postProcess(tree: Tree, copied: untpd.Tree): copied.ThisTree[T]
       protected def postProcess(tree: Tree, copied: untpd.MemberDef): copied.ThisTree[T]
 
+      /** Soucre of the copied tree */
+      protected def sourceFile(tree: Tree): SourceFile = tree.source
+
+      protected def sameSourceFile(tree: Tree): Boolean = tree.source == sourceFile(tree)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/Trees.scala]: Could be `true` for the default case.

File: compiler/src/dotty/tools/dotc/typer/Inliner.scala [hunk 2]
```diff
@@ -122,28 +120,48 @@ object Inliner {
...
-              case call :: _ if call.symbol.source != curSource =>
-                // Until we implement JSR-45, we cannot represent in output positions in other source files.
-                // So, reposition inlined code from other files with the call position:
-                transformed.withSpan(inlined.call.span)
-              case _ => transformed
-            }
+      class Reposition extends TreeMap(cpyWithNewSource) {
+        override def transform(tree: Tree)(implicit ctx: Context): Tree = {
+          val transformed = tree match {
+            case tree: Inlined => transformInline(tree)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Why are these treated separately?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala]: I believe it was to simplify the change of context while transforming. I will try to manually inline instead.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala]: I removed it

File: compiler/src/dotty/tools/dotc/ast/Trees.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Trees.scala]: Removed `sameSource`

File: compiler/src/dotty/tools/dotc/ast/Trees.scala [hunk 3]
```diff
@@ -990,187 +993,187 @@ object Trees {
...
-        case _ => finalize(tree, untpd.OrTypeTree(left, right)(tree.source))
+        case _ => finalize(tree, untpd.OrTypeTree(left, right)(sourceFile(tree)))
       }
       def RefinedTypeTree(tree: Tree)(tpt: Tree, refinements: List[Tree])(implicit ctx: Context): RefinedTypeTree = tree match {
         case tree: RefinedTypeTree if (tpt eq tree.tpt) && (refinements eq tree.refinements) => tree
-        case _ => finalize(tree, untpd.RefinedTypeTree(tpt, refinements)(tree.source))
+        case _ => finalize(tree, untpd.RefinedTypeTree(tpt, refinements)(sourceFile(tree)))
       }
       def AppliedTypeTree(tree: Tree)(tpt: Tree, args: List[Tree])(implicit ctx: Context): AppliedTypeTree = tree match {
         case tree: AppliedTypeTree if (tpt eq tree.tpt) && (args eq tree.args) => tree
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/ast/Trees.scala]: I remember from #5892 there used to be a check `sameSourceFile` in the case, why it is not needed anymore?

**Pull request merged**